frontend/utils.mjs

```diff
@@ -241,49 +240,48 @@ export async function buildOrWatch(config) {
         }
         buildsInProgress++
         onBuildStart?.(config)
-        reloadLiveServer()
         buildPromise = runBuild()
         const buildResponse = await buildPromise
         buildPromise = null
         await onBuildComplete?.(config, buildResponse)
         buildsInProgress--
         if (buildsInProgress === 0) {
             server?.resumeServer()
+            reloadLiveServer()
         }
+
         if (isDev && buildAgain) {
             void debouncedBuild()
         }
     }
 
-    let result = null
+    let esbuildContext = null
     let buildCount = 0
+    const activitySuccess = (count) => (count === 1 ? 'Built' : 'Rebuilt')
+    const activityError = (count) => (count === 1 ? 'Building failed' : 'Rebuilding failed ')
 
     async function runBuild() {
+        if (!esbuildContext) {
+            esbuildContext = await context({ ...commonConfig, ..._config })
+        }
+
         buildCount++
         const time = new Date()
-        if (buildCount === 1) {
-            console.log(`🧱 Building${name ? ` "${name}"` : ''}`)
-            try {
-                result = await build({ ...commonConfig, ..._config })
-                console.log(`🥇 Built${name ? ` "${name}"` : ''} in ${(new Date() - time) / 1000}s`)
-            } catch (error) {
-                console.log(`🛑 Building${name ? ` "${name}"` : ''} failed in ${(new Date() - time) / 1000}s`)
-                process.exit(1) // must exit since with result === null, result.rebuild() won't work
-            }
-        } else {
-            try {
-                result = await result.rebuild()
-                console.log(`🔄 Rebuilt${name ? ` "${name}"` : ''} in ${(new Date() - time) / 1000}s`)
-            } catch (e) {
-                console.log(`🛑 Rebuilding${name ? ` "${name}"` : ''} failed in ${(new Date() - time) / 1000}s`)
-            }
-        }
-        inputFiles = getInputFiles(result)
+        console.log(`🧱 ${name ? `"${name}": ` : ''}Building`)
+        try {
+            const buildResult = await esbuildContext.rebuild()
+            inputFiles = getInputFiles(buildResult)
 
-        return {
-            entrypoints: getBuiltEntryPoints(config, result),
-            chunks: getChunks(result),
-            ...result.metafile,
+            console.log(
+                `🥇 ${name ? `"${name}": ` : ''}${activitySuccess(buildCount)} in ${(new Date() - time) / 1000}s`
+            )
+            return {
+                entrypoints: getBuiltEntryPoints(config, buildResult),
+                chunks: getChunks(buildResult),
+                ...buildResult.metafile,
+            }
+        } catch (e) {
+            console.log(`🛑 ${name ? `"${name}": ` : ''}${activityError(buildCount)} in ${(new Date() - time) / 1000}s`)
         }
```

Comment:
I found it hard to see what was happening here given there was so much the same between the branches.

Since we no longer have to exit or call a different method the first time you could life logging logic out of here and it becomes if-less